addons/stock/models/product.py

```diff
@@ -833,12 +833,25 @@ def write(self, values):
                 lambda u: any(u[f].id != int(values[f]) if f in values else False
                               for f in {'category_id'}))
             if changed:
-                stock_move_lines = self.env['stock.move.line'].search_count([
-                    ('product_uom_id.category_id', 'in', changed.mapped('category_id.id')),
-                    ('state', '!=', 'cancel'),
+                mapped_ids = changed.mapped('id')
+                ok = False
+                stock_moves = self.env['stock.move'].search_count([
+                    ('product_uom', 'in', mapped_ids),
+                    ('state', 'not in', ('cancel', 'done')),
                 ])
-
-                if stock_move_lines:
+                if not stock_moves:
```

Comment:
I would rather use a structure like
```python3
error_msg = _( "You cannot change the ratio of this unit of mesure as some"
" products with this UoM have already been moved or are "
"currently reserved."
))
if self.env['stock.move'].search_count([
    ('product_uom', 'in', mapped_ids),
    ('state', 'not in', ('cancel', 'done')
)]):
    raise UserError(error_msg)
if self.env['stock.move.line'].search_count([
    ('product_uom_id', 'in', mapped_ids),
    ('state', 'not in', ('cancel', 'done')),
]):
    raise UserError(error_msg)
if self.env['stock.quant'].search_count([
    ('product_id.product_tmpl_id.uom_id', 'in', mapped_ids),
    ('quantity', '!=', 0),
]):
    raise UserError(error_msg)
```
I find it more readable

Reply:
:point_up_2: